working/static metaprogramming/feature_specification.md

```diff
@@ -0,0 +1,189 @@
+
+### Phase 3: Definition Macros
+
+The primary job of these macros is to fill in implementations of existing
+declarations (which must be abstract or external).
+
+They are also allowed to wrap existing methods or constructors, by injecting
+some code before and/or after those method bodies. These statements do share
+a scope with each other, but not with the original function body.
+
+In addition these macros can add supporting declarations to the surrounding
+scope, but these are private to the macro generated code, and never show up in
+introspection apis.
+
+These macros can fully introspect on any type reachable from the declarations
+they annotate, including introspecting on members of classes, etc.
+
+### Macro Ordering
+
+Macros are applied to inner declarations first - so for instance macros on class
+members are applied before macros on the class.
+
+When multiple macros are applied to the same declaration, they are applied right
+to left.
+
+Macros applied to different declarations in the same scope can be applied in any
+order.
+
+## APIs
+
+The specific apis for macros are being prototyped currently, and the docs are
+hosted [here][docs].
+
+### Macro API
+
+All macros implement a subtype of the [Macro][Macro] class, so you can follow
+the "Implementors" links to see all the available types of macros. Here are
+some direct links to the root interfaces for each phase:
+
+- [TypeMacro][TypeMacro]
+- [DeclarationMacro][DeclarationMacro]
+- [DefinitionMacro][DefinitionMacro]
+
+Each of these has more specific implementations for the specific type of
+declaration that they annotate. So for instance the
+[ClassDeclarationMacro][ClassDeclarationMacro] is the interface used for macros
+that apply to classes, and run in the "declaration" phase.
+
+### Introspection API
+
+The first argument to any method that you override in your macro is the
+introspection object. This is a representation of the declaration that was
+annotated with the macro.
+
+For example, lets look at the [ClassDeclarationMacro][ClassDeclarationMacro]
+API, which has the following method that you must override:
+
+```dart
+class ClassDeclarationMacro implements DeclarationMacro {
+  void visitClassDeclaration(
+      ClassDeclaration declaration, ClassDeclarationBuilder builder);
+}
+```
+
+The [ClassDeclaration][ClassDeclaration] instance you get here provides all the
+introspective information to you that is available for classes in the
+`declaration` phase.
+
+The `builder` instances also provide an api that allows you to retrieve an
+introspection object for any `Type` object available to the macro itself, so
+for instance the [DeclarationBuilder][DeclarationBuilder] class (which
+[ClassDeclarationBuilder][ClassDeclarationBuilder] extends), exposes the
+[typeDeclarationOf][typeDeclarationOf] api.
+
+### Code Building API
+
+At the root of the API for generating code, are the [Code][Code] and
+`*Builder` classes. The `*Builder` instance is always passed as the second
```

Comment:
How about inserting a sentence here like:

> The [Code][] class and its subtypes are first-class representations of pieces of Dart programs, essentially abstract syntax trees.

Reply:
Reworked this to include that, added a bullet for each of these after the first sentence